# Air Quality Dataset Aggregator

This notebook merges air quality data from multiple CSV files, filters by station ID, adds user-defined attributes, and splits the output into multiple files with a maximum of 10,000 rows each.

## Section 1: Import Required Libraries

In [15]:
import pandas as pd
import os
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')

## Section 2: Define User Input Parameters

Enter the station ID and the values for the new columns to be added to the dataset.

In [16]:
# Load the unified OSM features to get station data automatically
current_directory = os.getcwd()
unified_features_path = os.path.join(current_directory, "station_osm_features_unified.csv")

print("Loading unified OSM features...")
df_unified_features = pd.read_csv(unified_features_path)
print(f"✓ Loaded {len(df_unified_features)} stations from unified features")

# Identify all feature columns (exclude metadata columns)
metadata_cols = ['station_id', 'original_station_id', 'station_name', 
                 '_total_elements', '_unique_feature_types']
feature_columns = [col for col in df_unified_features.columns if col not in metadata_cols]

print(f"\nTotal feature columns: {len(feature_columns)}")
print(f"Feature columns: {feature_columns}")

print(f"\nFirst 5 stations (station info + first few features):")
display_cols = ['station_id', 'station_name'] + feature_columns[:5]
print(df_unified_features[display_cols].head())

print("\n✓ Ready to process all stations with ALL features automatically!")

Loading unified OSM features...
✓ Loaded 230 stations from unified features

Total feature columns: 28
Feature columns: ['latitude', 'longitude', 'aeroway', 'amenity', 'barrier', 'boundary', 'building', 'craft', 'emergency', 'healthcare', 'highway', 'historic', 'industrial', 'landuse', 'leisure', 'man_made', 'military', 'natural', 'office', 'place', 'power', 'public_transport', 'railway', 'shop', 'sport', 'tourism', 'water', 'waterway']

First 5 stations (station info + first few features):
  station_id                                     station_name  latitude  \
0      TN005                SIDCO Kurichi, Coimbatore - TNPCB     10.94   
1      KA012                    Urban, Chamarajanagar - KSPCB     11.55   
2      HR026                    MD University, Rohtak - HSPCB     28.52   
3      UP026  IESD Banaras Hindu University, Varanasi - UPPCB     25.26   
4      DL034                           Sirifort, Delhi - CPCB     28.55   

   longitude aeroway                                 

## Section 3: Load Station Hour Data

In [17]:
# Load station_hour.csv data
station_hour_path = os.path.join(current_directory, "station_hour.csv")

print(f"Loading station_hour.csv from: {station_hour_path}")
station_hour_df = pd.read_csv(station_hour_path)

print(f"✓ Station Hour DataFrame shape: {station_hour_df.shape}")
print(f"  Columns: {list(station_hour_df.columns)}")

Loading station_hour.csv from: /Users/likhithkanigolla/IIITH/code-files/Digital-Twin/Air-Quality-Dataset-Seggregator-DT/station_hour.csv
✓ Station Hour DataFrame shape: (2589083, 16)
  Columns: ['StationId', 'Datetime', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket']


In [18]:
#Drop Coloums 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'
cols_to_drop = ['NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket']
station_hour_df = station_hour_df.drop(columns=cols_to_drop)

In [19]:
# Print the first few rows of the station_hour_df to verify the data
print("\nFirst 5 rows of station_hour_df after dropping columns:")
print(station_hour_df.head())

print("\n" + "="*70)
print("DATASET INFO")
print("="*70)
print(station_hour_df.info())

# Print descriptive statistics (mean, std, min, max, etc.)
print("\n" + "="*70)
print("DESCRIPTIVE STATISTICS")
print("="*70)
# Suppress scientific notation for better readability
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')
print(station_hour_df.describe())

# Print statistics for PM2.5 and PM10 specifically
print("\n" + "="*70)
print("PM2.5 and PM10 STATISTICS")
print("="*70)
if 'PM2.5' in station_hour_df.columns and 'PM10' in station_hour_df.columns:
    print("\nPM2.5 Statistics:")
    print(f"  Mean: {station_hour_df['PM2.5'].mean():.2f}")
    print(f"  Median: {station_hour_df['PM2.5'].median():.2f}")
    print(f"  Std Dev: {station_hour_df['PM2.5'].std():.2f}")
    print(f"  Min: {station_hour_df['PM2.5'].min():.2f}")
    print(f"  Max: {station_hour_df['PM2.5'].max():.2f}")
    print(f"  Missing values: {station_hour_df['PM2.5'].isna().sum()}")
    
    print("\nPM10 Statistics:")
    print(f"  Mean: {station_hour_df['PM10'].mean():.2f}")
    print(f"  Median: {station_hour_df['PM10'].median():.2f}")
    print(f"  Std Dev: {station_hour_df['PM10'].std():.2f}")
    print(f"  Min: {station_hour_df['PM10'].min():.2f}")
    print(f"  Max: {station_hour_df['PM10'].max():.2f}")
    print(f"  Missing values: {station_hour_df['PM10'].isna().sum()}")
else:
    print("PM2.5 or PM10 columns not found in dataset")


First 5 rows of station_hour_df after dropping columns:
  StationId             Datetime  PM2.5   PM10
0     AP001  2017-11-24 17:00:00  60.50  98.00
1     AP001  2017-11-24 18:00:00  65.50 111.25
2     AP001  2017-11-24 19:00:00  80.00 132.00
3     AP001  2017-11-24 20:00:00  81.50 133.25
4     AP001  2017-11-24 21:00:00  75.25 116.00

DATASET INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2589083 entries, 0 to 2589082
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   StationId  object 
 1   Datetime   object 
 2   PM2.5      float64
 3   PM10       float64
dtypes: float64(2), object(2)
memory usage: 79.0+ MB
None

DESCRIPTIVE STATISTICS
             PM2.5         PM10
count 1,941,394.00 1,469,831.00
mean         80.86       158.48
std          89.48       139.79
min           0.01         0.01
25%          28.16        64.00
50%          52.59       116.25
75%          97.74       204.00
max       1,000.00     1,000.00

PM2.5 and PM10 STATI

## Section 4: Process All Stations Automatically

This section automatically processes each station from the unified features dataset.

In [20]:
# Process all stations automatically
print("=" * 70)
print("AUTOMATED STATION PROCESSING")
print("=" * 70)

# Check if StationId column exists
if 'StationId' not in station_hour_df.columns:
    print("✗ StationId not found in station_hour.csv")
    print(f"Available columns: {list(station_hour_df.columns)}")
else:
    print("✓ StationId found in station_hour.csv")

# Get list of stations to process (only those in station_hour data)
available_stations = station_hour_df['StationId'].unique()
stations_to_process = df_unified_features[df_unified_features['station_id'].isin(available_stations)]

print(f"\nTotal stations in unified features: {len(df_unified_features)}")
print(f"Stations with air quality data: {len(stations_to_process)}")
print(f"Stations to process: {len(stations_to_process)}")

if len(stations_to_process) == 0:
    print("\n⚠️ Warning: No stations to process!")
else:
    print(f"\n✓ Ready to process {len(stations_to_process)} stations")

AUTOMATED STATION PROCESSING
✓ StationId found in station_hour.csv

Total stations in unified features: 230
Stations with air quality data: 96
Stations to process: 96

✓ Ready to process 96 stations


In [21]:
# Check unique StationId values in station_hour.csv
unique_station_ids = station_hour_df['StationId'].unique()

print("=" * 70)
print("UNIQUE STATION IDs IN STATION_HOUR.CSV")
print("=" * 70)
print(f"Total unique stations: {len(unique_station_ids)}")
print(f"\nStation IDs:")
print(sorted(unique_station_ids))

UNIQUE STATION IDs IN STATION_HOUR.CSV
Total unique stations: 110

Station IDs:
['AP001', 'AP005', 'AS001', 'BR005', 'BR006', 'BR007', 'BR008', 'BR009', 'BR010', 'CH001', 'DL001', 'DL002', 'DL003', 'DL004', 'DL005', 'DL006', 'DL007', 'DL008', 'DL009', 'DL010', 'DL011', 'DL012', 'DL013', 'DL014', 'DL015', 'DL016', 'DL017', 'DL018', 'DL019', 'DL020', 'DL021', 'DL022', 'DL023', 'DL024', 'DL025', 'DL026', 'DL027', 'DL028', 'DL029', 'DL030', 'DL031', 'DL032', 'DL033', 'DL034', 'DL035', 'DL036', 'DL037', 'DL038', 'GJ001', 'HR011', 'HR012', 'HR013', 'HR014', 'JH001', 'KA002', 'KA003', 'KA004', 'KA005', 'KA006', 'KA007', 'KA008', 'KA009', 'KA010', 'KA011', 'KL002', 'KL004', 'KL007', 'KL008', 'MH005', 'MH006', 'MH007', 'MH008', 'MH009', 'MH010', 'MH011', 'MH012', 'MH013', 'MH014', 'ML001', 'MP001', 'MZ001', 'OD001', 'OD002', 'PB001', 'RJ004', 'RJ005', 'RJ006', 'TG001', 'TG002', 'TG003', 'TG004', 'TG005', 'TG006', 'TN001', 'TN002', 'TN003', 'TN004', 'TN005', 'UP012', 'UP013', 'UP014', 'UP015', '

## Filter Unified Features for Stations with Air Quality Data

Create a filtered version of the unified features containing only the 110 stations that have air quality data.

In [22]:
# Filter unified features to only include stations with air quality data
print("Filtering unified features for stations with air quality data...")
print("=" * 70)

# Filter the unified features dataframe
df_unified_filtered = df_unified_features[df_unified_features['station_id'].isin(unique_station_ids)].copy()

print(f"Original unified features: {len(df_unified_features)} stations")
print(f"Filtered unified features: {len(df_unified_filtered)} stations")
print(f"Stations excluded: {len(df_unified_features) - len(df_unified_filtered)} stations")

# Save the filtered dataset
output_file = 'station_osm_features_filtered_110.csv'
df_unified_filtered.to_csv(output_file, index=False)

print(f"\n✓ Saved filtered dataset to '{output_file}'")
print(f"  - Rows: {len(df_unified_filtered)}")
print(f"  - Columns: {len(df_unified_filtered.columns)}")

# Show some statistics
print("\n" + "-" * 70)
print("FILTERED DATASET SUMMARY")
print("-" * 70)
print(f"\nFirst 5 stations in filtered dataset:")
display_cols = ['station_id', 'station_name', 'latitude', 'longitude']
print(df_unified_filtered[display_cols].head())

print(f"\nAll {len(df_unified_filtered)} filtered station IDs:")
print(sorted(df_unified_filtered['station_id'].tolist()))

Filtering unified features for stations with air quality data...
Original unified features: 230 stations
Filtered unified features: 96 stations
Stations excluded: 134 stations

✓ Saved filtered dataset to 'station_osm_features_filtered_110.csv'
  - Rows: 96
  - Columns: 33

----------------------------------------------------------------------
FILTERED DATASET SUMMARY
----------------------------------------------------------------------

First 5 stations in filtered dataset:
   station_id                         station_name  latitude  longitude
0       TN005    SIDCO Kurichi, Coimbatore - TNPCB     10.94      76.98
4       DL034               Sirifort, Delhi - CPCB     28.55      77.22
10      DL017             Lodhi Road, Delhi - IITM     28.59      77.22
11      MH014                 Worli, Mumbai - MPCB     18.99      72.81
12      TN004  Velachery Res. Area, Chennai - CPCB     13.01      80.24

All 96 filtered station IDs:
['AP001', 'AP005', 'AS001', 'BR005', 'BR006', 'BR007', 'B

## Section 5: Process Each Station with Features

This section loops through each station, filters the data, adds OSM features, and saves to chunks.

In [23]:
# Loop through each station and process
chunk_size = 10000
total_stations = len(stations_to_process)
processed_count = 0
failed_count = 0
total_files_created = 0

# Target cleaning configuration (conservative bounds)
TARGET_POLLUTANTS = ['PM2.5', 'PM10']
PM25_RANGE = (0, 1000)  # µg/m³
PM10_RANGE = (0, 2000)  # µg/m³
OUTLIER_METHOD = 'winsorize'  # 'winsorize', 'iqr', 'none'
OUTLIER_LOWER_PCT = 0.01
OUTLIER_UPPER_PCT = 0.99

# Optional: drop other pollutant columns if present
OTHER_POLLUTANTS = ['NO', 'NO2', 'CO', 'CO2', 'SO2', 'O3', 'NH3', 'PM1', 'PM2_5', 'PM10_']

def clean_targets(df, target_cols):
    # Ensure numeric
    for col in target_cols:
        df[col] = pd.to_numeric(df[col], errors='coerce')
    # Drop rows with missing targets
    df = df.dropna(subset=target_cols).copy()
    # Apply conservative range clipping
    if 'PM2.5' in target_cols:
        df['PM2.5'] = df['PM2.5'].clip(lower=PM25_RANGE[0], upper=PM25_RANGE[1])
    if 'PM10' in target_cols:
        df['PM10'] = df['PM10'].clip(lower=PM10_RANGE[0], upper=PM10_RANGE[1])
    # Outlier handling
    if OUTLIER_METHOD == 'winsorize':
        for col in target_cols:
            lower = df[col].quantile(OUTLIER_LOWER_PCT)
            upper = df[col].quantile(OUTLIER_UPPER_PCT)
            df[col] = df[col].clip(lower=lower, upper=upper)
    elif OUTLIER_METHOD == 'iqr':
        for col in target_cols:
            q1 = df[col].quantile(0.25)
            q3 = df[col].quantile(0.75)
            iqr = q3 - q1
            lower = q1 - 1.5 * iqr
            upper = q3 + 1.5 * iqr
            df = df[(df[col] >= lower) & (df[col] <= upper)]
    return df

print(f"Starting automated processing of {total_stations} stations...")
print("=" * 70)

for idx, station_row in stations_to_process.iterrows():
    station_id = station_row['station_id']
    station_name = station_row['station_name']
    
    print(f"\n[{processed_count + 1}/{total_stations}] Processing: {station_id} - {station_name}")
    
    try:
        # Filter station_hour data for this station
        filtered_df = station_hour_df[station_hour_df['StationId'] == station_id].copy()
        
        if filtered_df.shape[0] == 0:
            print(f"  ⚠️ No data found for {station_id}, skipping...")
            failed_count += 1
            continue
        
        print(f"  Found {filtered_df.shape[0]} rows")
        
        # Keep only PM2.5 and PM10 targets (drop other pollutant columns if present)
        drop_cols = [col for col in OTHER_POLLUTANTS if col in filtered_df.columns and col not in TARGET_POLLUTANTS]
        if drop_cols:
            filtered_df = filtered_df.drop(columns=drop_cols)
        
        # Ensure target columns exist
        missing_targets = [t for t in TARGET_POLLUTANTS if t not in filtered_df.columns]
        if missing_targets:
            print(f"  ⚠️ Missing targets {missing_targets} for {station_id}, skipping...")
            failed_count += 1
            continue
        
        # Clean targets (ranges + outliers)
        before_rows = len(filtered_df)
        filtered_df = clean_targets(filtered_df, TARGET_POLLUTANTS)
        after_rows = len(filtered_df)
        print(f"  Cleaned targets: {before_rows} -> {after_rows} rows")
        if after_rows == 0:
            print(f"  ⚠️ No data left after cleaning for {station_id}, skipping...")
            failed_count += 1
            continue
        
        # Add ALL feature columns from unified dataset
        for feature_col in feature_columns:
            # Get the feature value for this station (handle NaN values)
            feature_value = station_row[feature_col]
            if pd.isna(feature_value):
                feature_value = "none"  # Use 'none' for missing features instead of empty string
            filtered_df[feature_col] = feature_value
        
        print(f"  Added {len(feature_columns)} feature columns")
        
        # Reorder columns: move PM2.5 and PM10 to the end for easy viewing
        # Get all columns except PM2.5 and PM10
        other_cols = [col for col in filtered_df.columns if col not in TARGET_POLLUTANTS]
        # Reorder: other columns first, then PM2.5 and PM10 at the end
        reordered_cols = other_cols + TARGET_POLLUTANTS
        filtered_df = filtered_df[reordered_cols]
        
        # Create output folder for this station
        output_folder = os.path.join(current_directory, "output", station_id)
        os.makedirs(output_folder, exist_ok=True)
        
        # Split into chunks and save
        total_rows = filtered_df.shape[0]
        num_chunks = (total_rows + chunk_size - 1) // chunk_size
        
        for i in range(num_chunks):
            start_idx = i * chunk_size
            end_idx = min((i + 1) * chunk_size, total_rows)
            chunk_df = filtered_df.iloc[start_idx:end_idx]
            
            file_number = i + 1
            filename = f"{station_id}_chunk_{file_number}.csv"
            filepath = os.path.join(output_folder, filename)
            
            chunk_df.to_csv(filepath, index=False)
            total_files_created += 1
        
        print(f"  ✓ Saved {num_chunks} chunk file(s) to output/{station_id}/")
        processed_count += 1
        
    except Exception as e:
        print(f"  ✗ Error processing {station_id}: {e}")
        failed_count += 1

# Final summary
print("\n" + "=" * 70)
print("PROCESSING COMPLETE!")
print("=" * 70)
print(f"Total stations processed: {processed_count}/{total_stations}")
print(f"Failed stations: {failed_count}")
print(f"Total files created: {total_files_created}")
print(f"\n✓ All output files saved in 'output/' folder")

Starting automated processing of 96 stations...

[1/96] Processing: TN005 - SIDCO Kurichi, Coimbatore - TNPCB
  Found 9229 rows
  Cleaned targets: 9229 -> 7503 rows
  Added 28 feature columns
  ✓ Saved 1 chunk file(s) to output/TN005/

[2/96] Processing: DL034 - Sirifort, Delhi - CPCB
  Found 32051 rows
  Cleaned targets: 32051 -> 18251 rows
  Added 28 feature columns
  ✓ Saved 2 chunk file(s) to output/DL034/

[3/96] Processing: DL017 - Lodhi Road, Delhi - IITM
  Found 24662 rows
  Cleaned targets: 24662 -> 21571 rows
  Added 28 feature columns
  ✓ Saved 3 chunk file(s) to output/DL017/

[4/96] Processing: MH014 - Worli, Mumbai - MPCB
  Found 9255 rows
  Cleaned targets: 9255 -> 8701 rows
  Added 28 feature columns
  ✓ Saved 1 chunk file(s) to output/MH014/

[5/96] Processing: TN004 - Velachery Res. Area, Chennai - CPCB
  Found 48192 rows
  Cleaned targets: 48192 -> 0 rows
  ⚠️ No data left after cleaning for TN004, skipping...

[5/96] Processing: TN002 - Manali Village, Chennai - TNP

## Summary

This automated notebook:
1. ✓ Loads the unified OSM features dataset
2. ✓ Automatically identifies ALL feature columns (not just 6)
3. ✓ Loads station_hour.csv with air quality data
4. ✓ Processes each station automatically:
   - Filters air quality data by station ID
   - Adds ALL OSM feature columns from unified dataset
   - Splits data into chunks of 10,000 rows
   - Saves chunks to `output/[StationID]/` folder
5. ✓ Provides detailed progress and summary statistics

All output files are organized in the `output/` folder by station ID!